board: nRF54L15DK support

kl-cruz opened the pull request:
The work comes with support for production board nRF54L15DK with device nRF54L15

kl-cruz committed:
dts: Add missing PDM entries to nRF54L15 device

Add PDM instances in nRF54L15.

Signed-off-by: Karol Lasończyk <karol.lasonczyk@nordicsemi.no>

kl-cruz committed:
soc: Add support for nRF54L15

Add support for production version of the device.

Signed-off-by: Karol Lasończyk <karol.lasonczyk@nordicsemi.no>

kl-cruz committed:
boards: nrf: Add nRF54L15 DK board

Add board with production version of the nRF54L15 SoC.

Signed-off-by: Karol Lasończyk <karol.lasonczyk@nordicsemi.no>

kl-cruz committed:
modules: Add support for nRF54L15 SoC

Update modules to support production version of the nRF54L15 SoC.

Signed-off-by: Karol Lasończyk <karol.lasonczyk@nordicsemi.no>

kl-cruz committed:
drivers: misc: Add support for nRF54L15 SoC

Expand vpr_launcher to support production version of the SoC.

Signed-off-by: Karol Lasończyk <karol.lasonczyk@nordicsemi.no>

nordicjm commented on boards/nordic/nrf54l15dk/board.cmake:
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+if (CONFIG_SOC_NRF54L15_CPUAPP)
+  board_runner_args(jlink "--device=cortex-m33" "--speed=4000")
+elseif (CONFIG_SOC_NRF54L15_CPUFLPR)
```
```suggestion
if(CONFIG_SOC_NRF54L15_CPUAPP)
  board_runner_args(jlink "--device=cortex-m33" "--speed=4000")
elseif(CONFIG_SOC_NRF54L15_CPUFLPR)
```

kl-cruz replied:
PDK configuration needs to be updated as well

nordicjm commented on boards/nordic/nrf54l15dk/doc/index.rst:
```diff
@@ -0,0 +1,138 @@
+Supported Features
+==================
+
+The ``nrf54l15dk/nrf54l15/cpuapp`` board configuration supports the following
```
board target

kl-cruz replied:
PDK doc needs to be updated as well

nordicjm commented on boards/nordic/nrf54l15dk/doc/index.rst:
```diff
@@ -0,0 +1,138 @@
++-----------+------------+----------------------+
+| GPIO      | on-chip    | gpio                 |
++-----------+------------+----------------------+
+| TWIM      | on-chip    | i2c                  |
```
no twis support?

kl-cruz replied:
Only on nrfx level. No implementation in i2c shim

nordicjm commented on boards/nordic/nrf54l15dk/doc/index.rst:
```diff
@@ -0,0 +1,138 @@
+Programming and Debugging
+*************************
+
+Applications for the ``nrf54l15dk/nrf54l15/cpuapp`` board can be
```
board target

nordicjm commented on boards/nordic/nrf54l15dk/nrf54l15dk_nrf54l15_cpuapp_defconfig:
```diff
@@ -0,0 +1,29 @@
+# Enable hardware stack protection
+CONFIG_HW_STACK_PROTECTION=y
+
+# MPU-based null-pointer dereferencing detection cannot
+# be applied as the (0x0 - 0x400) is unmapped for this target.
+CONFIG_NULL_POINTER_EXCEPTION_DETECTION_NONE=y
```
why can't this be selected at soc-level?

kl-cruz replied:
It is based on PDK configuration

kl-cruz replied:
I would prefer to do this update on both configuration at once

kl-cruz committed:
test: samples: nRF54L15 support

Commit consists of minimal set of overlays need to run tests
and samples. Additionally disables nRF54L15 platform in cpp98
tests, because unnamed unions are not supported.

Signed-off-by: Karol Lasończyk <karol.lasonczyk@nordicsemi.no>

kl-cruz committed:
samples: drivers: mbox: Update mbox sample to work with nRF54L15

Update mbox sample configuration to work with production
board and SoC.

Signed-off-by: Karol Lasończyk <karol.lasonczyk@nordicsemi.no>

kl-cruz committed:
tests: kernel: Add support for FLPR core in nRF54L15

Add support for production device.

Signed-off-by: Karol Lasończyk <karol.lasonczyk@nordicsemi.no>

kl-cruz committed:
dts: boards: Add nRF54L15 ENGA configuration

Add conditional DTS compilation in case of ENGA version.

Signed-off-by: Karol Lasończyk <karol.lasonczyk@nordicsemi.no>

kl-cruz commented:
Rebased on top of: https://github.com/zephyrproject-rtos/zephyr/pull/77160

kl-cruz commented:
> Rebased on top of: #77160

PR merged, rebased on top of main

gmarull commented on soc/nordic/nrf54l/Kconfig.defconfig.nrf54l15_cpuapp:
```diff
@@ -0,0 +1,14 @@
+config NUM_IRQS
+	default 271
+
+config IEEE802154_NRF5
+	default IEEE802154
```
is this needed?

e-rk replied:
It looks like other SoCs use this option. I would leave it as is for this PR and I will check if it's needed for this and other SoCs afterwards.

gmarull replied:
should not be needed, option is already default y based on DT. Just remove it and other socs can be updated later.

kl-cruz replied:
removed

gmarull commented on soc/nordic/nrf54l/Kconfig.defconfig.nrf54l15_cpuflpr:
```diff
@@ -0,0 +1,15 @@
+
+if SOC_NRF54L15_CPUFLPR
+
+config RISCV_HAS_CPU_IDLE
+	bool
```
this should be selected somewhere else?

kl-cruz replied:
See no references - removed

gmarull commented on soc/nordic/nrf54l/Kconfig.defconfig.nrf54l15_cpuflpr:
```diff
@@ -0,0 +1,15 @@
+config RISCV_HAS_CPU_IDLE
+	bool
+
+config NUM_IRQS
+	int
```
`int` not needed, this is just defaulting

kl-cruz replied:
done

gmarull commented on boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr.dts:
```diff
@@ -21,6 +21,21 @@
+
+&cpuflpr_vevif_tx {
+	nordic,events-mask = <0x00020000>;
+};
```
why is board modifying this?

kl-cruz replied:
should be ok now

kl-cruz replied:
I do not see it in https://github.com/zephyrproject-rtos/zephyr/pull/77341/files#diff-8e93c65c9c86c18f332d6de70f274bda09e6a0fea3110153c93d0bfb251fbad8

gmarull replied:
ping, I still see this in board files. question is why soc level config is kept in a board file?

kl-cruz replied:
done

kl-cruz commented on boards/nordic/nrf54l15pdk/nrf54l15_cpuapp_common.dtsi:
```diff
@@ -22,6 +22,14 @@
+
+&cpuapp_vevif_tx {
+	nordic,tasks-mask = <0x0003f800>;
+};
```
should be ok now

jaz1-nordic commented on boards/nordic/nrf54l15pdk/nrf54l15_cpuapp_common.dtsi:
```diff
@@ -6,7 +6,7 @@
 /* This file is common to the secure and non-secure domain */
 
-#include <nordic/nrf54l15_cpuapp.dtsi>
+#include <nordic/nrf54l15_cpuapp_enga.dtsi>
```
```suggestion
#include <nordic/nrf54l15_enga_cpuapp.dtsi>
```
To be compatible with Kconfigs `CONFIG_SOC_NRF54L15_ENGA_CPUAPP`?

kl-cruz replied:
done

jaz1-nordic commented on boards/nordic/nrf54l15pdk/nrf54l15pdk_nrf54l15_cpuflpr.dts:
```diff
@@ -5,7 +5,7 @@
 
 /dts-v1/;
-#include <nordic/nrf54l15_cpuflpr.dtsi>
+#include <nordic/nrf54l15_cpuflpr_enga.dtsi>
```
Same here?

kl-cruz replied:
done

nordicjm commented on boards/nordic/nrf54l15dk/doc/index.rst:
```diff
@@ -0,0 +1,138 @@
+| WDT       | on-chip    | watchdog             |
++-----------+------------+----------------------+
+
+Other hardware features have not been enabled yet for this board.
```
seems like no documentation for flpr

kl-cruz replied:
Should I add new entry for flpr in this table? like:
FLPR | on-chip | arch/riscv

nordicjm replied:
Not in the table, but information on how to build for that core

kl-cruz replied:
working on it

nordicjm commented on boards/nordic/nrf54l15dk/nrf54l15_cpuapp_common.dtsi:
```diff
@@ -0,0 +1,156 @@
+
+	mx25r64: mx25r6435f@0 {
+		compatible = "jedec,spi-nor";
+		status = "disabled";
```
why is this disabled?

kl-cruz replied:
Copied from reference board L15PDK

nordicjm replied:
I would say the l15pdk is wrong then, doesn't really make sense to add this and disable it, no other board does that

kl-cruz replied:
I can enable it by default. Unless it comes with some implications

hubertmis replied:
The implication would be power, startup time, and memory overhead required for the driver.
Not sure if the power overhead is only during initialization or also run-time. It depends on the proper driver support for the power management.

I guess the intention was to simply enable external memory in overlays of samples utilizing it (simple `status = "okay";`) to avoid this overhead in most samples.

nordicjm replied:
This would only be enabled by default if an application enables FLASH, so for things like MCUboot, and there should be a configuration for MCUboot that disables it if it's not needed (as per nrf52840dk/nrf5340dk), additional overhead in power/start up time/memory for tests is negligible

nordicjm reviewed (dismissed):
Can either update PR with documentation on how to use flpr or add it in another PR, either/or but please make sure it is added

gmarull commented on dts/arm/nordic/nrf54l15_cpuapp.dtsi:
```diff
@@ -44,15 +49,23 @@ nvic: &cpuapp_nvic {};
 		interrupts = <76 NRF_DEFAULT_IRQ_PRIORITY>;
 		#mbox-cells = <1>;
 		nordic,events = <1>;
+#ifdef NRF54L15_ENGA
 		nordic,events-mask = <0x00008000>;
+#else
+		nordic,events-mask = <0x00100000>;
+#endif
```
does this work? where is this define injected?

why not simply have a common file for enga/non-enga, and a specific file with the necessary overrides for each?

kl-cruz replied:
done by override

gmarull commented on dts/arm/nordic/nrf54l15_enga_cpuapp.dtsi:
```diff
@@ -0,0 +1,9 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define NRF54L15_ENGA 1
```
I see the file.

Please, avoid using ifdeffery in dts files, they support overloading

gmarull replied:
not solved

gmarull commented on snippets/nordic-flpr/boards/nrf54l15dk_nrf54l15_cpuapp.overlay:
```diff
@@ -0,0 +1,43 @@
+	};
+};
+
+&uart30 {
+	status = "reserved";
+};
+
+&cpuapp_sram {
+	reg = <0x20000000 DT_SIZE_K(160)>;
+	ranges = <0x0 0x20000000 0x28000>;
+};
+
+&cpuflpr_vpr {
+	execution-memory = <&cpuflpr_sram_code_data>;
+	source-memory = <&cpuflpr_code_partition>;
+};
+
+&cpuapp_vevif_tx {
```
this is a copy of PDK file, please, use snippets facility to include using regexes, `l15*dk`. cc: @nordicjm

carlescufi replied:
Given that the PDK will go away, this is a minor issue?

gmarull replied:
There's a trivial solution: use globbing in the snippet, so please, update.

gmarull reviewed (dismissed):
snippets part can be improved later

kl-cruz commented:
Errors not related to my changes
